### DISCUSSION TITLE
How to set a parameter in onCreateQueryBuilder

### DISCUSSION THREAD

**Discussion by devsrealm** [pimcore/pimcore#15741] [Q&A / Support]
Why is this not working:

```
...
$queryBuilder
->andWhere('u_prod=.slug = :slug')
->setParameter(':slug', $slug);
```

**Comment by NiklasBr** [comment]
No colon in `setParameter`:

```php
$queryBuilder
    ->andWhere('table.myField = :myNamedParameter')
    ->setParameter('myNamedParameter', $varaible);
```

**Reply by devsrealm** [reply]
That did not work either, figured out a new way by querying u_prod directly and filtering by the returned object

**Reply by NiklasBr** [reply]
It does work, maybe you did not show the entire context? For example if the above is _exactly_ the query you are trying to run, then yes it will not work not only because of the parameters you are asking about, but also because you are not defining which table to select from.

For a full example:

```php
$queryBuilder
    ->select('*')
    ->from('some_table', 'table')
    ->where('table.o_published = 1')
    ->andWhere('table.myField = :myNamedParameter')
    ->setParameter('myNamedParameter', $varaible);
```

**Reply by devsrealm** [reply]
You are right, I did not show the full query, here is an example of what I am trying to do:

```php
$products = Product::getList([
            "order" => "desc",
            "orderKey" => "o_id",
            "published" => true
        ]);

  $products->onCreateQueryBuilder(
            function (QueryBuilder $queryBuilder) use ($products) {

                $table = $products->getDao()->getTableName();
                $uTable = "object_localized_u_prod_{$this->getRequest()->getLocale()}";

                $queryBuilder->leftJoin($table, $uTable, 'u_prod',"$table.uProd__id = u_prod.o_id")
->andWhere('u_prod.slug = :slug')
->setParameter('slug', $slug);

            }
        );
```

Aside from the fact that it did not work, though the joint statement worked, it failed to populate the `u_prod` property that is used to link the table, I think that is another question entirely though (it populated it fine previously so I don't know what went wrong).

**Reply by NiklasBr** [reply]
Aha, yes, as part of `onCreateQueryBuilder` with a Listing it gets more finicky, I have also had problems there, could be a mix of named and indexed parameters. Maybe `"published" => true` creates an indexed param which does not work well with the slug one?

**Reply by devsrealm** [reply]
Yeah, really finicky as you noticed, even with not having any config in the `getList()` method, e.g: `$products = OnlineShopProduct::getList();` it doesn't work, went with a hack that works fine for now.

**Reply by kovinet** [reply]
I also noticed this doesn't work.  I am using normal string concatenation like 
```php
$queryBuilder->where('documents.path LIKE "' . $pathPrefix . '%"');
``` 
since I was not able to find better solution.
